inception-project/inception
# #4491 - Support for link feature recommendations

**What's in the PR**
- Allow recommenders to generate suggestions for link features
- Added unit tests

**How to test manually**
* Set up a project with a `webanno.custom.Arg` layer
* Set up a project with a `webanno.custom.Span` layer
* Add a link feature from Span to Arg
* Set up an external recommender with the following predict function
```
    def predict(self, cas: Cas, layer: str, feature: str, project_id: str, document_id: str, user_id: str):
        Sentence = cas.typesystem.get_type("de.tudarmstadt.ukp.dkpro.core.api.segmentation.type.Sentence")
        LinkHost = cas.typesystem.get_type(layer)
        LinkType = cas.typesystem.get_type(layer+"ArgsLink")
        SlotFiller = cas.typesystem.get_type("webanno.custom.Arg")
        FSArray = cas.typesystem.get_type(TYPE_NAME_FS_ARRAY)

        canidateHosts = cas.select(LinkHost)
        for candidateHost in canidateHosts:
            sentence = next(cas.select_covering(Sentence, candidateHost))
            slotFiller = cas.select_covered(SlotFiller, sentence)[0]
            prediction = create_prediction(cas, layer, feature, candidateHost.get("begin"), candidateHost.get("end"))
            prediction.set(feature, FSArray(elements=[LinkType(role="blah", target=slotFiller)]))
            cas.add(prediction)
```
* Create a Span and an Arg annotation in the same sentence
* Note that a link suggestion between span and Arg should be generated
* Click on the suggestion to accept it

<img width="293" alt="Screenshot 2024-02-04 at 17 31 23" src="https://github.com/inception-project/inception/assets/1410238/f28e20a4-aee7-41bc-9ce3-f657e4ba4e59">

**Automatic testing**

**Documentation**
